### ISSUE TITLE
Some triangles are completely and some partially ignored

### ISSUE THREAD

**Issue by EmJay276** [jacquemv/evenlyspacedstreamlines#1]
Hi, 

this is a really awesome package you built!
But I face some issues with triangles being completely (see blue circle) or partially (see green circles) ignored / not filled with lines, even if enough space is available.

Do you have any idea how to fix this?

My settings are basically default except for a seed
```python
# create streamlines
(list_of_lines_original,
 list_of_triangle_indices,
 infos) = evenly_spaced_streamlines(triangle_vertices,
                                    triangle_indices,
                                    fiber_orientation,
                                    radius,
                                    seed_points=32,  # default: 32
                                    max_angle=90,  # default: 90
                                    avoid_u_turns=True,  # default: True
                                    random_seed=1698460061)  # default: 0 -> based on clock
```

Mesh
<img src="https://github.com/user-attachments/assets/54617f75-62c2-4118-9701-2e50fa498138"/>

Resulting streamlines
<img src="https://github.com/user-attachments/assets/320a0581-e179-4060-9efe-10e61702e673"/>

**Comment by jacquemv** [comment]
Hello,

Thanks for your interest!

It's caused by a limitation of the seeding process. It currently does not allow the first segment of a streamline to be smaller than a triangle (it has to touch two edges).
In this case, finding an open space in a triangle is a 1D search (implemented using a 1D occupancy mask). Being able to place smaller segments anywhere in a triangle would require a 2D search which is more computationally expensive. In my application, I was removing the very small streamlines anyway, so it wasn't a big deal for me.

I've thought about that in the past and I think it could be done as an optional refinement step, going triangle by triangle and looking for small gaps.
In summary, there is no easy fix, but it motivates me to look into this as soon as I have time.

Best regards,
Vincent

De : Michael Jäger ***@***.***>
Envoyé : 11 décembre 2024 09:57
À : jacquemv/evenlyspacedstreamlines ***@***.***>
Cc : Subscribed ***@***.***>
Objet : [jacquemv/evenlyspacedstreamlines] Some triangles are complete and some partially ignored (Issue #1)


Hi,

this is a really awesome package you built!
But I face some issues with triangles being completely (see blue circle) or partially (see green circles) ignored / not filled with lines, even if enough space is available.

Do you have any idea how to fix this?

My settings are basically default except for a seed

# create streamlines

(list_of_lines_original,

 list_of_triangle_indices,

 infos) = evenly_spaced_streamlines(triangle_vertices,

                                    triangle_indices,

                                    fiber_orientation,

                                    radius,

                                    seed_points=32,  # default: 32

                                    max_angle=90,  # default: 90

                                    avoid_u_turns=True,  # default: True

                                    random_seed=1698460061)  # default: 0 -> based on clock

Mesh
Mesh.png (view on web)<https://github.com/user-attachments/assets/54617f75-62c2-4118-9701-2e50fa498138>

Resulting streamlines
evenlyspacedstreamlines_marked.png (view on web)<https://github.com/user-attachments/assets/320a0581-e179-4060-9efe-10e61702e673>

-

**Comment by EmJay276** [comment]
Thank you very much for your very fast reply!

I also remove very small lines (length < 2 * radius), but my triangles are compared to the radius very large (the triangle edge length is at least  5-10x radius).

Would it be a possible workaround to remesh my triangle to a much smaller size so only very small "gaps" are created (which I don't care about)? Or will this slow down the agorithm a lot?

**Comment by jacquemv** [comment]
Yes, the most important parameter is the ratio between the radius and triangle edge length.
You can remesh and interpolate the vector field. The optimum tends to be when triangle edge length is about equal to the radius.
When the radius is much smaller or much larger, the algorithm is slower.

De : Michael Jäger ***@***.***>
Envoyé : 11 décembre 2024 11:57
À : jacquemv/evenlyspacedstreamlines ***@***.***>
Cc : Vincent Jacquemet ***@***.***>; Comment ***@***.***>
Objet : Re: [jacquemv/evenlyspacedstreamlines] Some triangles are complete and some partially ignored (Issue #1)


Thank you very much for your very fast reply!

I also remove very small lines (length < 2 * radius), but my triangles are compared to the radius very large (the triangle edge length is at least 5-10x radius).

Would it be a possible workaround to remesh my triangle to a much smaller size so only very small "gaps" are created (which I don't care about)? Or will this slow down the agorithm a lot?

-

**Comment by EmJay276** [comment]
A finer and more uniformed mesh worked really well, thank you very much!
Is there any option to enforce "longer" lines or to enforce some connection at the ends of two almost coaxial lines?

**Comment by EmJay276** [comment]
Here is an example (the red lines are your algorithm, yellow is my subsequent shift of them)
![image](https://github.com/user-attachments/assets/95634174-30e4-4f8d-921b-aa01e52157f7)

**Comment by jacquemv** [comment]
I haven't found a way to fix this without creating problems elsewhere. Naturally, the streamline stops even if it becomes a tiny bit too close to other streamlines.
I think it's best to use a postprocessing step in this case.
Make a loop that looks for streamline endpoints that are at a distance < radius * 1.5 or something like that and concatenate the streamlines.
Maybe another condition could be the angles of the new connecting edge shouldn't be too large.
But if the vector field has singularities, it can be tricky to find the right trade-off.

De : Michael Jäger ***@***.***>
Envoyé : 19 décembre 2024 07:52
À : jacquemv/evenlyspacedstreamlines ***@***.***>
Cc : Vincent Jacquemet ***@***.***>; Comment ***@***.***>
Objet : Re: [jacquemv/evenlyspacedstreamlines] Some triangles are completely and some partially ignored (Issue #1)


Here is an example (the red lines are your algorithm, yellow is my subsequent shift of them)
image.png (view on web)<https://github.com/user-attachments/assets/95634174-30e4-4f8d-921b-aa01e52157f7>

-

**Comment by EmJay276** [comment]
Thanks for your info! Postprocessing the streamlines and connecting the "best-matching ones" recursively worked really well (better than expected). But it was a bit of a hassle to get it performant.

For anyone reading this in the future, for me this gave the best results: 
- Removing streamlines with a total length less than 2x radius
- Then merging streamlines recursively
   - Condition: Ends within a 1.5x radius 
   - Starting from the "most matching" lines (cosine distance closest to -1; with each direction vector pointing from each end "inwards":  A-->----<--B;   C-->------<--D; Lines A-B and C-D with cos_dist(vec(B), vec(C)) = -1
   - Repeat until no more line pairs can be merged
- I did not need to limit my maximum cosine distance. 
   - A value of 1 worked still fine for me, introducing U-turns (which I don't care about) but increasing the mean line width significantly.
   - Limiting the max cosine distance to 0 results in a kind of "better looking" result without changing the overall appearance

**Comment by jacquemv** [comment]
Great!

De : Michael Jäger ***@***.***>
Envoyé : 22 décembre 2024 16:25
À : jacquemv/evenlyspacedstreamlines ***@***.***>
Cc : Vincent Jacquemet ***@***.***>; Comment ***@***.***>
Objet : Re: [jacquemv/evenlyspacedstreamlines] Some triangles are completely and some partially ignored (Issue #1)


Thanks for your info! Postprocessing the streamlines and connecting the "best-matching ones" recursively worked really well (better than expected). But it was a bit of a hassle to get it performant.

For anyone reading this in the future, for me this gave the best results:

  *   Removing streamlines with a total length less than 2x radius
  *   Then merging streamlines recursively
     *   Condition: Ends within a 1.5x radius
     *   Starting from the "most matching" lines (cosine distance closest to -1; with each direction vector pointing from each end "inwards": A-->----<--B; C-->------<--D; Lines A-B and C-D with cos_dist(vec(B), vec(C)) = -1
     *   Repeat until no more line pairs can be merged
  *   I did not need to limit my maximum cosine distance.
     *   A value of 1 worked still fine for me, introducing U-turns (which I don't care about) but increasing the mean line width significantly.
     *   Limiting the max cosine distance to 0 results in a kind of "better looking" result without changing the overall appearance

-
